File: Android/wallet/src/main/java/io/outblock/wallet/storage/StorageProtocol.kt
package io.outblock.wallet.storage

import io.outblock.wallet.errors.WalletError

/**
 * Protocol defining storage behavior for wallet data
 */
interface StorageProtocol {
    /**
     * Get all keys currently stored in the storage
     */
    val allKeys: List<String>

    /**
     * Find keys matching a keyword
     * @param keyword Search term to match against keys
     * @return List of matching keys
     * @throws WalletError if operation fails
     */
    fun findKey(keyword: String): List<String>

    /**
     * Get data from storage
     * @param key The key to retrieve data for
     * @return The stored data, or null if not found
     * @throws WalletError if operation fails
     */
    fun get(key: String): ByteArray?

    /**
     * Store data in storage
     * @param key The key to store data under
     * @param data The data to store
     * @throws WalletError if operation fails
     */
    fun set(key: String, data: ByteArray)

    /**
     * Remove data from storage
     * @param key The key to remove data for
     * @throws WalletError if operation fails
     */
    fun remove(key: String)

    /**
     * Remove all data from storage
     * @throws WalletError if operation fails
     */
    fun removeAll()
}

/**
 * Enum representing different security levels for storage backends
 */
enum class SecurityLevel {
    /**
     * Highest security level - uses hardware-backed encryption
     * (e.g., Android Keystore, Secure Enclave)
     */
    HARDWARE_BACKED,

    /**
     * Medium security level - uses standard secure storage
     * (e.g., SharedPreferences with MODE_PRIVATE)
     */
    STANDARD,

    /**
     * Lowest security level - uses in-memory storage
     * Data is lost when app is closed
     */
    IN_MEMORY
} 

File: typescript/src/storage/StorageProtocol.ts
/**
 * Flow Wallet Kit - Storage Protocol Base Implementation
 * 
 * This module provides a base implementation of the StorageProtocol interface
 * that can be extended by concrete storage providers.
 */

import { StorageProtocol, SecurityLevel } from '../types/storage.js';
import { WalletError, WalletErrorCode } from '../utils/errors.js';

/**
 * Abstract base class for storage implementations
 * Provides common functionality and validation
 */
export abstract class BaseStorageProtocol implements StorageProtocol {
  /**
   * Internal storage map for keys
   */
  protected abstract readonly storage: Map<string, Uint8Array>;
  
  /**
   * Security level of the storage implementation
   */
  public abstract readonly securityLevel: SecurityLevel;
  
  /**
   * Get all keys currently stored
   */
  public get allKeys(): readonly string[] {
    return Array.from(this.storage.keys());
  }
  
  /**
   * Find keys matching a keyword
   * @param keyword - Search term to match against keys
   * @returns List of matching keys
   */
  public async findKey(keyword: string): Promise<readonly string[]> {
    try {
      if (!keyword) {
        return [];
      }
      
      const lowerKeyword = keyword.toLowerCase();
      const matchingKeys = this.allKeys.filter(key => 
        key.toLowerCase().includes(lowerKeyword)
      );
      
      return matchingKeys;
    } catch (error) {
      throw new WalletError(
        WalletErrorCode.LoadCacheFailed,
        `Failed to find keys: ${error instanceof Error ? error.message : 'Unknown error'}`,
        error
      );
    }
  }
  
  /**
   * Get data from storage
   * @param key - The key to retrieve data for
   * @returns The stored data, or null if not found
   */
  public abstract get(key: string): Promise<Uint8Array | null>;
  
  /**
   * Store data in storage
   * @param key - The key to store data under
   * @param data - The data to store
   */
  public abstract set(key: string, data: Uint8Array): Promise<void>;
  
  /**
   * Remove data from storage
   * @param key - The key to remove data for
   */
  public abstract remove(key: string): Promise<void>;
  
  /**
   * Remove all data from storage
   */
  public abstract removeAll(): Promise<void>;
  
  /**
   * Validate storage key
   * @param key - Key to validate
   * @throws {WalletError} if key is invalid
   */
  protected validateKey(key: string): void {
    if (!key || typeof key !== 'string') {
      throw new WalletError(
        WalletErrorCode.EmptyKey,
        'Storage key must be a non-empty string'
      );
    }
    
    if (key.length > 255) {
      throw new WalletError(
        WalletErrorCode.EmptyKey,
        'Storage key must not exceed 255 characters'
      );
    }
  }
  
  /**
   * Validate storage data
   * @param data - Data to validate
   * @throws {WalletError} if data is invalid
   */
  protected validateData(data: Uint8Array): void {
    if (!(data instanceof Uint8Array)) {
      throw new WalletError(
        WalletErrorCode.EmptyKey,
        'Storage data must be a Uint8Array'
      );
    }
    
    if (data.length === 0) {
      throw new WalletError(
        WalletErrorCode.EmptyKey,
        'Storage data must not be empty'
      );
    }
  }
}